File: 2020/day03.py
from functools import partial, reduce
from operator import mul

with open("inputs/day03.input", "r") as f:
    grid = list(line.strip() for line in f.readlines())


def trees(grid, r, d):
    return sum(row[(i*r) % len(row)] == "#"
               for i, row in enumerate(grid[::d]))


part1 = partial(trees, r=3, d=1)


def part2(grid):
    skips = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
    return reduce(mul, (trees(grid, *skip) for skip in skips))


print(f"Part 1: {part1(grid)}")
print(f"Part 2: {part2(grid)}")

File: rust-2020/src/day03.rs
pub fn read(contents: &str) -> Vec<&str> {
    contents.lines().collect()
}

fn trees(grid: &[&str], row: usize, col: usize) -> usize {
    let ncol = grid[0].len();
    grid.iter()
        .step_by(col)
        .enumerate()
        .filter(|(i, line)| line.chars().nth(i * row % ncol).unwrap() == '#')
        .count()
}

pub fn part1(data: &[&str]) -> usize {
    trees(data, 3, 1)
}

pub fn part2(data: &[&str]) -> usize {
    let skips = vec![(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)];
    skips
        .iter()
        .map(|(row, col)| trees(data, *row, *col))
        .product()
}

pub fn run() -> (String, String) {
    let contents = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day03.input"));
    let data = read(contents);
    let p1 = part1(&data).to_string();
    let p2 = part2(&data).to_string();
    (p1, p2)
}
